### PR TITLE
[GOB-1027] Add Import from GOB API functionality

### PR DESCRIPTION
**Pull request by barykine**

### CODE REVIEW

File: src/.flake8 [hunk 1]
```diff
@@ -13,5 +13,5 @@ exclude=.git,
         tests,
         scripts,
         */settings_commons.py
-max-complexity=6
+max-complexity=7
```
Comment (Vegter) [inline src/.flake8]: This is not "allowed", well we do never do this.

Comment (barykine) [reply src/.flake8]: the complexity is in _flatten_edge which is a copy from GOB-Export which has no complexity check. then we have to refactor _flatten_edge (which I tried to avoid) or not to use it.

File: src/gobprepare/csv_stream.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
+class CSVStream():
```
Comment (JKoetsier) [inline src/gobprepare/csv_stream.py]: This class would be a good candidate to move in a separate GOB-Lib or something. Really generic code, and rather complex. Moving this into a separate lib repo can prevent trouble later on.

File: src/gobprepare/importers/api_importer.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+      description
+      type {
+        name
+      }
+    }
+  }
+}
+''' % meta_type
+
+    def get_meta_fields(self, meta_data):
```
Comment (JKoetsier) [inline src/gobprepare/importers/api_importer.py]: Nice! :)

File: src/gobprepare/prepare_client.py [hunk 4]
```diff
@@ -37,6 +40,7 @@ class PrepareClient:
     _src_user = None
     _dst_connection = None
     _dst_user = None
+    _dst_raw_connection = None
```
Comment (JKoetsier) [inline src/gobprepare/prepare_client.py]: Never used.

File: src/gobprepare/prepare_client.py [hunk 5]
```diff
@@ -70,6 +74,7 @@ def connect(self):
         """
         self._src_connection, self._src_user = self._connect_application(self.source)
         self._dst_connection, self._dst_user = self._connect_application(self.destination)
+        self._dst_engine = self._create_engine(self.destination)
```
Comment (JKoetsier) [inline src/gobprepare/prepare_client.py]: Why use two different connection methods? Rather confusing. PostgresAPIImporter should use self._dst_connection as well.

File: src/gobprepare/prepare_client.py [hunk 6]
```diff
@@ -183,27 +200,54 @@ def action_import_csv(self, action: dict):
...
         elif src == "file":
-            with open(action["query"]) as f:
-                return f.read()
+            return self._read_file(action[source_name])
+
+        raise GOBException(f"Missing or invalid '{source_type}'")
 
-        raise GOBException("Missing or invalid 'query_src'")
+    def _get_query(self, action: dict):
+        return self._get_query_per_source_type(action, 'query_src', 'query')
```
Comment (JKoetsier) [inline src/gobprepare/prepare_client.py]: Why this extra complexity? _get_query_per_source_type is never used otherwise. Seems unnecessary and adds to the complexity in understanding the code.

File: src/gobprepare/sql.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    :return:
+    """
+    return f"""
+-- DROP SCHEMA {name} CASCADE;
+CREATE SCHEMA IF NOT EXISTS {_quote(name)};
+COMMIT;
+"""
+
+
+def get_create_table_sql(schema, table_name, description, meta_fields, field_names):
```
Comment (JKoetsier) [inline src/gobprepare/sql.py]: Similar code exists in the Prepare cloner. Combine both? Same goes for the quote function above. Very similar code in the same repo.

File: src/gobprepare/requests.py [hunk 8]
```diff
@@ -0,0 +1,15 @@
+import requests
```
Comment (JKoetsier) [inline src/gobprepare/requests.py]: All these files (requests, sql, csv, graphql, graphql_stream, etc) are put in the root of gobprepare. I'd put everything in a different module to keep the root of the gobprepare module clean and understandable. (If I were a new developer on the project, I should be directed to the code that is the core of the repo, the prepare_client. Auxiliary code such as request, sql, csv handling aren't important and should be put in their own namespace, or grouped in an api_importer module as long as api_importer is the only module using that code.)

File: src/gobprepare/utils/graphql.py [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    @staticmethod
+    def format_item(item):
+        # Convert to list
+        items = item if isinstance(item, list) else [item]
+        yield from items
+
+    def __iter__(self):
+        items = post_stream(f'{self.host}{self.endpoint}', {'query': self.query})
+        for item in items:
+            yield from self.format_item(json.loads(item))
```
[the remark this answers is not in the document]
Comment (Vegter) [reply src/gobprepare/utils/graphql.py]:
```
yield from [self.format_item(json.loads(item)) for item in items]
```

File: src/gobprepare/importers/api_importer.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+WRITE_BATCH_SIZE = 10000
+
+
+def get_host():
+    """API Host
+
+    :return: The API host to get the collection from
+    """
+    return os.getenv('API_HOST', 'http://localhost:8141')
+
```
[the remark this answers is not in the document]
Comment (barykine) [reply src/gobprepare/importers/api_importer.py]: done

File: src/gobprepare/importers/api_importer.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+    fields {
+      name
+      description
+      type {
+        name
+      }
+    }
+  }
+}
+''' % meta_type
```
[the remark this answers is not in the document]
Comment (barykine) [reply src/gobprepare/importers/api_importer.py]: wontfix

File: src/gobprepare/importers/api_importer.py [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+            if counter >= WRITE_BATCH_SIZE:
+                self.write_rows(rows)
+                rows = []
+                counter = 0
+        self.write_rows(rows)
+
+    @staticmethod
+    def get_row(node, field_names):
+        return [node[field_name] for field_name in field_names]
+
```
[the remark this answers is not in the document]
Comment (barykine) [reply src/gobprepare/importers/api_importer.py]: renamed to get_row_values as it is about values, not field names

File: src/gobprepare/utils/graphql.py [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+class GraphQL:
+
+    def __init__(self, host, endpoint, query):
+        self.host = host
+        self.endpoint = endpoint
+        self.query = query
+
+    @staticmethod
+    def format_item(item):
+        # Convert to list
```
[the remark this answers is not in the document]
Comment (barykine) [reply src/gobprepare/utils/graphql.py]: done

File: src/gobprepare/importers/api_importer.py [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+}
+''' % meta_type
+
+    def get_meta_fields(self, meta_data):
+        meta_fields = {}
+        for meta in meta_data:
+            for field in meta["data"]["__type"]["fields"]:
+                meta_fields[field["name"]] = field
+        return meta_fields
+
```
[the remark this answers is not in the document]
Comment (barykine) [reply src/gobprepare/importers/api_importer.py]: done

File: src/gobprepare/config.py [hunk 15]
```diff
@@ -8,6 +8,11 @@
 POSTGRES_DRIVER = 'postgresql'
 DEBUG = os.getenv('GOB_DEBUG', False)
 
+STREAMING_GRAPHQL_ENDPOINT = '/gob/graphql/streaming/'
+GRAPHQL_ENDPOINT = '/gob/graphql/'
+WRITE_BATCH_SIZE = 10000
+DEFAULT_API_HOST = 'http://localhost:8141'
+
```
Comment (Vegter) [inline src/gobprepare/config.py]: Why not set GOB_API_HOST here so that all config (including read from env) is located at one place.

File: src/gobprepare/importers/api_importer.py [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+            meta_fields.update({field["name"]: field for field in meta["data"]["__type"]["fields"]})
+        return meta_fields
+
+    def get_field_names(self, entities):
+        """Get field names from a first data record.
+
+        :param entities: Data entities.
+        :return: List of field names.
+        """
+        first_record = next(entities)
```
Comment (Vegter) [inline src/gobprepare/importers/api_importer.py]: What happens if no data can be read (GOB is still empty). Having GOB data is a prerequisite to this import so it should either be asserted with a logical error description or otherwise be handled. It should also be documented in the README

File: src/gobprepare/utils/sql.py [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+        'type': SQL_TYPE_CONVERSIONS[type],
+        'description': description
+    }
+
+
+def get_create_table_sql(schema, table_name, description, meta_fields, field_names):
+    """
+    Returns a SQL statement to create a table in a schema
+    The table fields are constructed from GraphQL meta fields
+
```
Comment (Vegter) [inline src/gobprepare/utils/sql.py]: The first field is the primary key. Is this somewhere related to the format or order of the GraphQL query? If so then this should be documented with the respective query.

File: src/tests/gobprepare/importers/test_api_importer.py [hunk 18]
```diff
@@ -0,0 +1,178 @@
+import pytest
+from unittest import TestCase
+from unittest.mock import patch, call, ANY
+
+from gobprepare.config import (
```
Comment (Vegter) [inline src/tests/gobprepare/importers/test_api_importer.py]: Can you add a test for an empty GOB, ie GOB API returns an empty list?

File: src/gobprepare/utils/sql.py (hunk 17 above)
Comment (JKoetsier) [reply src/gobprepare/utils/sql.py]: A primary key isn't even necessary. Logic to decide on primary key is based on implicit knowledge and assumptions. Why not just forget about the primary key? This way it can never go wrong, and keeps the code much simpler.

File: src/tests/gobprepare/importers/test_api_importer.py (hunk 18 above)
Comment (barykine) [reply src/tests/gobprepare/importers/test_api_importer.py]: goog one. will do.

File: src/gobprepare/config.py (hunk 15 above)
Comment (barykine) [reply src/gobprepare/config.py]: will do.

Comment (barykine) [reply src/gobprepare/config.py]: done

File: src/gobprepare/utils/sql.py (hunk 17 above)
Comment (barykine) [reply src/gobprepare/utils/sql.py]: fixed

**Comment by Vegter** [comment]
Is there a reason to force push every time? If individual commits get all squashed into one commit it is no longer possible to just check the changes. It makes it a little hard to review the code because every time you have to go through all code changes, instead of just the changes due to the last comments.

**Comment by barykine** [comment]
totally agree. we can always squash at the end on approval just before merge.

**Comment by Vegter** [comment]
Within Datapunt we have guidelines for commit messages. You can find an example in the hall on our floor. If you follow these guidelines you will end up with logically complete and well documented commits. Within GOB we try to follow these rules and review comments on commits and the documentation of commits are very welcome! So please have a look at the Datapunt guidelines. Thanks

File: src/gobprepare/utils/graphql.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Vegter) [reply src/gobprepare/utils/graphql.py]: This one is still open

File: src/gobprepare/importers/api_importer.py (hunk 16 above)
Comment (Vegter) [reply src/gobprepare/importers/api_importer.py]: Can you add this as a requirement to the README. Prepare can only run when this precondition is met

File: src/gobprepare/utils/graphql.py [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+    @staticmethod
+    def item_to_list(item):
+        # Convert to list
+        items = item if isinstance(item, list) else [item]
+        yield from items
+
+    def __iter__(self):
+        items = post_stream(f'{self.host}{self.endpoint}', {'query': self.query})
+        for item in items:
+            yield from self.item_to_list(json.loads(item))
```
Comment (barykine) [inline src/gobprepare/utils/graphql.py]: renamed the method.

File: src/gobprepare/utils/graphql.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Vegter) [reply src/gobprepare/utils/graphql.py]: You can probably add a simple example in the comments

File: src/gobprepare/utils/graphql.py (hunk 19 above)
Comment (Vegter) [reply src/gobprepare/utils/graphql.py]: That's right but it still is difficult to understand what happens. Why is it an array of arrays? Just comment an example of what is received of describe what happens should be enough I think.
